finat/fiat_elements.py

```diff
@@ -173,6 +176,119 @@ def point_evaluation(self, order, refcoords, entity=None):
         # Dispatch on FIAT element class
         return point_evaluation(self._element, order, refcoords, (entity_dim, entity_i))
 
+    @property
+    def dual_basis(self):
+        # A tensor of weights (of total rank R) to contract with a unique
+        # vector of points to evaluate at, giving a tensor (of total rank R-1)
+        # where the first indices (rows) correspond to a basis functional
+        # (node).
+        # DOK Sparse matrix in (row, col, higher,..)=>value pairs - to become a
+        # gem.SparseLiteral.
+        # Rows are number of nodes/dual functionals.
+        # Columns are unique points to evaluate.
+        # Higher indices are tensor indices of the weights when weights are
+        # tensor valued.
+        # The columns (unique points to evaluate) are indexed out ready for
+        # contraction with the point set free index.
+        Q = {}
+
+        # Dict of unique points to evaluate stored as
+        # {tuple(pt_hash.flatten()): (x_k, k)} pairs. Points in index k
+        # order form a vector required for correct contraction with Q. Will
+        # become a FInAT.PointSet later.
+        # x_key = numpy.round(x_k, x_key_decimals) such that
+        # numpy.allclose(x_k, pt_hash, atol=1*10**-dec) == true
+        x = {}
+        x_key_decimals = 12
+        x_key_atol = 1e-12  # = 1*10**-dec
+
+        #
+        # BUILD Q TENSOR
+        #
+
+        # FIXME: The below loop is REALLY SLOW for BDM - Q and x should just be output as the dual basis
+
+        last_shape = None
+        self.Q_is_identity = True  # TODO do this better
+
+        fiat_dual_basis = self._element.dual_basis()
+        # i are rows of Q
+        for i, dual in enumerate(fiat_dual_basis):
+            point_dict = dual.get_point_dict()
+            for j, (x_j, tuples) in enumerate(point_dict.items()):
+                x_j = PointSet((x_j,))  # TODO doesn't need to be PointSet
+                # get weights q_j for point x_j from complicated FIAT dual data
+                # structure
+                weight_dict = {cmp: weight for weight, cmp in tuples}
+                if len(self._element.value_shape()) == 0:
+                    q_j = np.array(weight_dict[tuple()])
+                else:
+                    q_j = np.zeros(self._element.value_shape())
+                    for key, item in weight_dict.items():
+                        q_j[key] = item
+
+                # Ensure all weights have the same shape
+                if last_shape is not None:
+                    assert q_j.shape == last_shape
+                last_shape = q_j.shape
+
+                # Create key into x dict
+                x_key = np.round(x_j.points, x_key_decimals)
+                assert np.allclose(x_j.points, x_key, atol=x_key_atol)
+                x_key = tuple(x_key.flatten())
```

Comment:
I originally did this because there were cases where the same point was being treated as not-the-same-point to within some small error. This could of course have been because of my initial wrapping in a `PointSet`. I'll test if your re-implementation without rounding still causes any issues (sadly I didn't not the exact test case). I guess even if it does it's probably safer not to force points to be the same within some magical tolerance